KT-20635 - Kotlin plugin extension with a method accepting Action<in TYPE> method does not get inferred as SAM

### Expected Behavior

* `Action<in TYPE>` should behave same way in Groovy and Kotlin plugins
* script authors should not have to cast lambdas for `Action` methods
* Kotlin plugin authors should have clear documentation for either extension functions (`().Type -> Unit`) or using `Action<>` types for best behavior across both Groovy and Kotlin build scripts

### Current Behavior

Running `./gradlew -s --scan` with example project (steps shown at bottom to reproduce) results in exception for type mismatch.

<details><summary>Build output</summary><p>

```
e: /home/mkobit/Workspace/personal/mkobit/jenkins-pipeline-shared-library-example/build.gradle.kts:60:22: Type mismatch: inferred type is () -> Unit but Action<in PluginDependencySpec> was expected

FAILURE: Build failed with an exception.

* What went wrong:
A problem occurred configuring root project 'jenkins-pipeline-shared-library-example'.
> Could not open cache directory 6z7hjh1lgsmu0365r759yuli3 (/home/mkobit/.gradle/caches/4.2.1/gradle-kotlin-dsl/6z7hjh1lgsmu0365r759yuli3).
   > Internal error: unable to compile script, see log for details

* Try:
Run with --info or --debug option to get more log output.

* Exception is:
org.gradle.api.ProjectConfigurationException: A problem occurred configuring root project 'jenkins-pipeline-shared-library-example'.
	at org.gradle.configuration.project.LifecycleProjectEvaluator.addConfigurationFailure(LifecycleProjectEvaluator.java:94)
	at org.gradle.configuration.project.LifecycleProjectEvaluator.doConfigure(LifecycleProjectEvaluator.java:66)
	at org.gradle.configuration.project.LifecycleProjectEvaluator.access$100(LifecycleProjectEvaluator.java:34)
	at org.gradle.configuration.project.LifecycleProjectEvaluator$ConfigureProject.run(LifecycleProjectEvaluator.java:110)
	at org.gradle.internal.progress.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:336)
	at org.gradle.internal.progress.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:328)
	at org.gradle.internal.progress.DefaultBuildOperationExecutor.execute(DefaultBuildOperationExecutor.java:199)
	at org.gradle.internal.progress.DefaultBuildOperationExecutor.run(DefaultBuildOperationExecutor.java:110)
	at org.gradle.configuration.project.LifecycleProjectEvaluator.evaluate(LifecycleProjectEvaluator.java:50)
	at org.gradle.api.internal.project.DefaultProject.evaluate(DefaultProject.java:656)
	at org.gradle.api.internal.project.DefaultProject.evaluate(DefaultProject.java:135)
	at org.gradle.execution.TaskPathProjectEvaluator.configure(TaskPathProjectEvaluator.java:35)
	at org.gradle.execution.TaskPathProjectEvaluator.configureHierarchy(TaskPathProjectEvaluator.java:60)
	at org.gradle.configuration.DefaultBuildConfigurer.configure(DefaultBuildConfigurer.java:38)
	at org.gradle.initialization.DefaultGradleLauncher$ConfigureBuild.run(DefaultGradleLauncher.java:249)
	at org.gradle.internal.progress.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:336)
	at org.gradle.internal.progress.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:328)
	at org.gradle.internal.progress.DefaultBuildOperationExecutor.execute(DefaultBuildOperationExecutor.java:199)
	at org.gradle.internal.progress.DefaultBuildOperationExecutor.run(DefaultBuildOperationExecutor.java:110)
	at org.gradle.initialization.DefaultGradleLauncher.configureBuild(DefaultGradleLauncher.java:167)
	at org.gradle.initialization.DefaultGradleLauncher.doBuildStages(DefaultGradleLauncher.java:126)
	at org.gradle.initialization.DefaultGradleLauncher.executeTasks(DefaultGradleLauncher.java:109)
	at org.gradle.internal.invocation.GradleBuildController$1.call(GradleBuildController.java:78)
	at org.gradle.internal.invocation.GradleBuildController$1.call(GradleBuildController.java:75)
	at org.gradle.internal.work.DefaultWorkerLeaseService.withLocks(DefaultWorkerLeaseService.java:152)
	at org.gradle.internal.invocation.GradleBuildController.doBuild(GradleBuildController.java:100)
	at org.gradle.internal.invocation.GradleBuildController.run(GradleBuildController.java:75)
	at org.gradle.tooling.internal.provider.ExecuteBuildActionRunner.run(ExecuteBuildActionRunner.java:28)
	at org.gradle.launcher.exec.ChainingBuildActionRunner.run(ChainingBuildActionRunner.java:35)
	at org.gradle.tooling.internal.provider.ValidatingBuildActionRunner.run(ValidatingBuildActionRunner.java:32)
	at org.gradle.launcher.exec.RunAsBuildOperationBuildActionRunner$1.run(RunAsBuildOperationBuildActionRunner.java:43)
	at org.gradle.internal.progress.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:336)
	at org.gradle.internal.progress.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:328)
	at org.gradle.internal.progress.DefaultBuildOperationExecutor.execute(DefaultBuildOperationExecutor.java:199)
	at org.gradle.internal.progress.DefaultBuildOperationExecutor.run(DefaultBuildOperationExecutor.java:110)
	at org.gradle.launcher.exec.RunAsBuildOperationBuildActionRunner.run(RunAsBuildOperationBuildActionRunner.java:40)
	at org.gradle.tooling.internal.provider.SubscribableBuildActionRunner.run(SubscribableBuildActionRunner.java:51)
	at org.gradle.launcher.exec.InProcessBuildActionExecuter.execute(InProcessBuildActionExecuter.java:47)
	at org.gradle.launcher.exec.InProcessBuildActionExecuter.execute(InProcessBuildActionExecuter.java:30)
	at org.gradle.launcher.exec.BuildTreeScopeBuildActionExecuter.execute(BuildTreeScopeBuildActionExecuter.java:39)
	at org.gradle.launcher.exec.BuildTreeScopeBuildActionExecuter.execute(BuildTreeScopeBuildActionExecuter.java:25)
	at org.gradle.tooling.internal.provider.ContinuousBuildActionExecuter.execute(ContinuousBuildActionExecuter.java:80)
	at org.gradle.tooling.internal.provider.ContinuousBuildActionExecuter.execute(ContinuousBuildActionExecuter.java:53)
	at org.gradle.tooling.internal.provider.ServicesSetupBuildActionExecuter.execute(ServicesSetupBuildActionExecuter.java:57)
	at org.gradle.tooling.internal.provider.ServicesSetupBuildActionExecuter.execute(ServicesSetupBuildActionExecuter.java:32)
	at org.gradle.tooling.internal.provider.GradleThreadBuildActionExecuter.execute(GradleThreadBuildActionExecuter.java:36)
	at org.gradle.tooling.internal.provider.GradleThreadBuildActionExecuter.execute(GradleThreadBuildActionExecuter.java:25)
	at org.gradle.tooling.internal.provider.ParallelismConfigurationBuildActionExecuter.execute(ParallelismConfigurationBuildActionExecuter.java:43)
	at org.gradle.tooling.internal.provider.ParallelismConfigurationBuildActionExecuter.execute(ParallelismConfigurationBuildActionExecuter.java:29)
	at org.gradle.tooling.internal.provider.StartParamsValidatingActionExecuter.execute(StartParamsValidatingActionExecuter.java:64)
	at org.gradle.tooling.internal.provider.StartParamsValidatingActionExecuter.execute(StartParamsValidatingActionExecuter.java:29)
	at org.gradle.tooling.internal.provider.SessionFailureReportingActionExecuter.execute(SessionFailureReportingActionExecuter.java:59)
	at org.gradle.tooling.internal.provider.SessionFailureReportingActionExecuter.execute(SessionFailureReportingActionExecuter.java:44)
	at org.gradle.tooling.internal.provider.SetupLoggingActionExecuter.execute(SetupLoggingActionExecuter.java:45)
	at org.gradle.tooling.internal.provider.SetupLoggingActionExecuter.execute(SetupLoggingActionExecuter.java:30)
	at org.gradle.launcher.daemon.server.exec.ExecuteBuild.doBuild(ExecuteBuild.java:67)
	at org.gradle.launcher.daemon.server.exec.BuildCommandOnly.execute(BuildCommandOnly.java:36)
	at org.gradle.launcher.daemon.server.api.DaemonCommandExecution.proceed(DaemonCommandExecution.java:120)
	at org.gradle.launcher.daemon.server.exec.WatchForDisconnection.execute(WatchForDisconnection.java:37)
	at org.gradle.launcher.daemon.server.api.DaemonCommandExecution.proceed(DaemonCommandExecution.java:120)
	at org.gradle.launcher.daemon.server.exec.ResetDeprecationLogger.execute(ResetDeprecationLogger.java:26)
	at org.gradle.launcher.daemon.server.api.DaemonCommandExecution.proceed(DaemonCommandExecution.java:120)
	at org.gradle.launcher.daemon.server.exec.RequestStopIfSingleUsedDaemon.execute(RequestStopIfSingleUsedDaemon.java:34)
	at org.gradle.launcher.daemon.server.api.DaemonCommandExecution.proceed(DaemonCommandExecution.java:120)
	at org.gradle.launcher.daemon.server.exec.ForwardClientInput$2.call(ForwardClientInput.java:74)
	at org.gradle.launcher.daemon.server.exec.ForwardClientInput$2.call(ForwardClientInput.java:72)
	at org.gradle.util.Swapper.swap(Swapper.java:38)
	at org.gradle.launcher.daemon.server.exec.ForwardClientInput.execute(ForwardClientInput.java:72)
	at org.gradle.launcher.daemon.server.api.DaemonCommandExecution.proceed(DaemonCommandExecution.java:120)
	at org.gradle.launcher.daemon.server.exec.LogAndCheckHealth.execute(LogAndCheckHealth.java:55)
	at org.gradle.launcher.daemon.server.api.DaemonCommandExecution.proceed(DaemonCommandExecution.java:120)
	at org.gradle.launcher.daemon.server.exec.LogToClient.doBuild(LogToClient.java:62)
	at org.gradle.launcher.daemon.server.exec.BuildCommandOnly.execute(BuildCommandOnly.java:36)
	at org.gradle.launcher.daemon.server.api.DaemonCommandExecution.proceed(DaemonCommandExecution.java:120)
	at org.gradle.launcher.daemon.server.exec.EstablishBuildEnvironment.doBuild(EstablishBuildEnvironment.java:82)
	at org.gradle.launcher.daemon.server.exec.BuildCommandOnly.execute(BuildCommandOnly.java:36)
	at org.gradle.launcher.daemon.server.api.DaemonCommandExecution.proceed(DaemonCommandExecution.java:120)
	at org.gradle.launcher.daemon.server.exec.StartBuildOrRespondWithBusy$1.run(StartBuildOrRespondWithBusy.java:50)
	at org.gradle.launcher.daemon.server.DaemonStateCoordinator$1.run(DaemonStateCoordinator.java:295)
	at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:63)
	at org.gradle.internal.concurrent.ManagedExecutorImpl$1.run(ManagedExecutorImpl.java:46)
	at org.gradle.internal.concurrent.ThreadFactoryImpl$ManagedThreadRunnable.run(ThreadFactoryImpl.java:55)
Caused by: org.gradle.cache.CacheOpenException: Could not open cache directory 6z7hjh1lgsmu0365r759yuli3 (/home/mkobit/.gradle/caches/4.2.1/gradle-kotlin-dsl/6z7hjh1lgsmu0365r759yuli3).
	at org.gradle.cache.internal.DefaultPersistentDirectoryStore.open(DefaultPersistentDirectoryStore.java:61)
	at org.gradle.cache.internal.DefaultPersistentDirectoryStore.open(DefaultPersistentDirectoryStore.java:32)
	at org.gradle.cache.internal.DefaultCacheFactory.doOpen(DefaultCacheFactory.java:90)
	at org.gradle.cache.internal.DefaultCacheFactory.open(DefaultCacheFactory.java:64)
	at org.gradle.cache.internal.DefaultCacheRepository$PersistentCacheBuilder.open(DefaultCacheRepository.java:123)
	at org.gradle.kotlin.dsl.cache.ScriptCache.cacheDirFor(ScriptCache.kt:51)
	at org.gradle.kotlin.dsl.cache.ScriptCache.cacheDirFor$default(ScriptCache.kt:43)
	at org.gradle.kotlin.dsl.provider.CachingKotlinCompiler.cacheDirFor(CachingKotlinCompiler.kt:170)
	at org.gradle.kotlin.dsl.provider.CachingKotlinCompiler.compileScript(CachingKotlinCompiler.kt:131)
	at org.gradle.kotlin.dsl.provider.CachingKotlinCompiler.compileBuildScript(CachingKotlinCompiler.kt:111)
	at org.gradle.kotlin.dsl.provider.KotlinBuildScriptCompiler.compileScriptFile(KotlinBuildScriptCompiler.kt:242)
	at org.gradle.kotlin.dsl.provider.KotlinBuildScriptCompiler.executeScriptBodyOn(KotlinBuildScriptCompiler.kt:114)
	at org.gradle.kotlin.dsl.provider.KotlinBuildScriptCompiler.prepareAndExecuteScriptBodyOn(KotlinBuildScriptCompiler.kt:108)
	at org.gradle.kotlin.dsl.provider.KotlinBuildScriptCompiler.access$prepareAndExecuteScriptBodyOn(KotlinBuildScriptCompiler.kt:51)
	at org.gradle.kotlin.dsl.provider.KotlinBuildScriptCompiler$compileTopLevelScript$1.invoke(KotlinBuildScriptCompiler.kt:91)
	at org.gradle.kotlin.dsl.provider.KotlinBuildScriptCompiler$compileTopLevelScript$1.invoke(KotlinBuildScriptCompiler.kt:51)
	at org.gradle.kotlin.dsl.provider.KotlinScriptPlugin.apply(KotlinScriptPlugin.kt:48)
	at org.gradle.configuration.BuildOperationScriptPlugin$1.run(BuildOperationScriptPlugin.java:61)
	at org.gradle.internal.progress.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:336)
	at org.gradle.internal.progress.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:328)
	at org.gradle.internal.progress.DefaultBuildOperationExecutor.execute(DefaultBuildOperationExecutor.java:199)
	at org.gradle.internal.progress.DefaultBuildOperationExecutor.run(DefaultBuildOperationExecutor.java:110)
	at org.gradle.configuration.BuildOperationScriptPlugin.apply(BuildOperationScriptPlugin.java:58)
	at org.gradle.configuration.project.BuildScriptProcessor.execute(BuildScriptProcessor.java:41)
	at org.gradle.configuration.project.BuildScriptProcessor.execute(BuildScriptProcessor.java:26)
	at org.gradle.configuration.project.ConfigureActionsProjectEvaluator.evaluate(ConfigureActionsProjectEvaluator.java:34)
	at org.gradle.configuration.project.LifecycleProjectEvaluator.doConfigure(LifecycleProjectEvaluator.java:64)
	... 80 more
Caused by: java.lang.IllegalStateException: Internal error: unable to compile script, see log for details
	at org.gradle.kotlin.dsl.support.KotlinCompilerKt.compileKotlinScriptToDirectory(KotlinCompiler.kt:84)
	at org.gradle.kotlin.dsl.provider.CachingKotlinCompiler$compileScriptTo$$inlined$run$lambda$1.invoke(CachingKotlinCompiler.kt:155)
	at org.gradle.kotlin.dsl.provider.CachingKotlinCompiler$compileScriptTo$$inlined$run$lambda$1.invoke(CachingKotlinCompiler.kt:47)
	at org.gradle.kotlin.dsl.provider.CachingKotlinCompiler.withProgressLoggingFor(CachingKotlinCompiler.kt:216)
	at org.gradle.kotlin.dsl.provider.CachingKotlinCompiler.compileScriptTo(CachingKotlinCompiler.kt:153)
	at org.gradle.kotlin.dsl.provider.CachingKotlinCompiler.access$compileScriptTo(CachingKotlinCompiler.kt:47)
	at org.gradle.kotlin.dsl.provider.CachingKotlinCompiler$compileScript$cacheDir$1.invoke(CachingKotlinCompiler.kt:133)
	at org.gradle.kotlin.dsl.provider.CachingKotlinCompiler$compileScript$cacheDir$1.invoke(CachingKotlinCompiler.kt:47)
	at org.gradle.kotlin.dsl.cache.ScriptCacheKt$sam$Action$904ef041.execute(ScriptCache.kt)
	at org.gradle.cache.internal.DefaultPersistentDirectoryCache$Initializer.initialize(DefaultPersistentDirectoryCache.java:106)
	at org.gradle.cache.internal.FixedSharedModeCrossProcessCacheAccess$1.run(FixedSharedModeCrossProcessCacheAccess.java:85)
	at org.gradle.cache.internal.DefaultFileLockManager$DefaultFileLock.doWriteAction(DefaultFileLockManager.java:196)
	at org.gradle.cache.internal.DefaultFileLockManager$DefaultFileLock.writeFile(DefaultFileLockManager.java:186)
	at org.gradle.cache.internal.FixedSharedModeCrossProcessCacheAccess.open(FixedSharedModeCrossProcessCacheAccess.java:83)
	at org.gradle.cache.internal.DefaultCacheAccess.open(DefaultCacheAccess.java:141)
	at org.gradle.cache.internal.DefaultPersistentDirectoryStore.open(DefaultPersistentDirectoryStore.java:59)
	... 106 more


* Get more help at https://help.gradle.org

BUILD FAILED in 2s

The build scan was not published due to a configuration problem.

The Gradle Cloud Services license agreement has not been agreed to.

To agree to the license, include the following in your root project's configuration:
buildScan { licenseAgreementUrl = 'https://gradle.com/terms-of-service'; licenseAgree = 'yes' }

For more information, please see https://gradle.com/scans/help/plugin-license.

Alternatively, if you are using Gradle Enterprise, specify the server location.
For more information, please see https://gradle.com/scans/help/plugin-enterprise-config.
```

### Context

Writing a plugin in Kotlin and wanted to provided a friendly DSL for both Groovy and Kotlin.
I believe that in order to get the friendly DSL in both Groovy and Kotlin an `Action<? super TYPE>` method and ` TYPE.() -> Unit` methods have to be added (not 100% on that).

### Steps to Reproduce (for bugs)

* See `Action {` cast at https://github.com/mkobit/jenkins-pipeline-shared-library-example/blob/master/build.gradle.kts#L60
* If the `Action` casting is removed, get above stack trace
* See https://github.com/mkobit/jenkins-pipeline-shared-library-example/blob/62a65e3b20dda56430ef2c10acfa885b92f4849d/build.gradle.kts#L59 for method definition of `fun pluginDependencies(action: Action<in PluginDependencySpec>) { //... }`

</p></details>

### Your Environment

* Gradle 4.2.1
 * Build scan URL: 

*NOTE*: Cannot build scan in this fashion with invalid a compiler failure in build script

--- a/subprojects/plugins/src/main/kotlin/org/gradle/kotlin/dsl/plugins/dsl/KotlinDslCompilerPlugins.kt
+++ b/subprojects/plugins/src/main/kotlin/org/gradle/kotlin/dsl/plugins/dsl/KotlinDslCompilerPlugins.kt
@@ -20,6 +20,10 @@ import org.gradle.api.HasImplicitReceiver
 import org.gradle.api.Plugin
 import org.gradle.api.Project
 
+import org.gradle.kotlin.dsl.withType
+
+import org.jetbrains.kotlin.gradle.tasks.KotlinCompile
+
 import org.jetbrains.kotlin.samWithReceiver.gradle.SamWithReceiverExtension
 import org.jetbrains.kotlin.samWithReceiver.gradle.SamWithReceiverGradleSubplugin
 
@@ -36,5 +40,19 @@ open class KotlinDslCompilerPlugins : Plugin<Project> {
         extensions.configure(SamWithReceiverExtension::class.java) { samWithReceiver ->
             samWithReceiver.annotation(HasImplicitReceiver::class.qualifiedName!!)
         }
+
+        tasks.withType<KotlinCompile> {
+            enableSamConversionForKotlinFunctions()
+        }
+    }
+}
+
+
+fun KotlinCompile.enableSamConversionForKotlinFunctions() {
+    kotlinOptions {
+        freeCompilerArgs += listOf(
+            "-XXLanguage:+NewInference",
+            "-XXLanguage:+SamConversionForKotlinFunctions"
+        )
     }
 }

--- a/subprojects/plugins/src/test/kotlin/org/gradle/kotlin/dsl/plugins/dsl/KotlinDslPluginTest.kt
+++ b/subprojects/plugins/src/test/kotlin/org/gradle/kotlin/dsl/plugins/dsl/KotlinDslPluginTest.kt
@@ -3,6 +3,7 @@ package org.gradle.kotlin.dsl.plugins.dsl
 import org.gradle.kotlin.dsl.fixtures.customDaemonRegistry
 import org.gradle.kotlin.dsl.fixtures.customInstallation
 import org.gradle.kotlin.dsl.fixtures.AbstractPluginTest
+import org.gradle.kotlin.dsl.fixtures.containsMultiLineString
 
 import org.gradle.testkit.runner.TaskOutcome
 
@@ -225,6 +226,61 @@ class KotlinDslPluginTest : AbstractPluginTest() {
         val result = buildWithPlugin("classes")
 
         assertThat(result.outcomeOf(":compileKotlin"), equalTo(TaskOutcome.SUCCESS))
+    }
+
+    @Test
+    fun `enables SAM conversion for Kotlin functions`() {
+
+        withSettingsIn("buildSrc", pluginManagementBlock)
+
+        withBuildScriptIn("buildSrc", """
+
+            plugins {
+                `kotlin-dsl`
+            }
+
+        """)
+
+        withFile("buildSrc/src/main/kotlin/my.kt", """
+            package my
+
+            // Action<T> is a SAM with receiver
+            fun <T> applyActionTo(value: T, action: org.gradle.api.Action<T>) = action.execute(value)
+
+            // NamedDomainObjectFactory<T> is a regular SAM
+            fun <T> create(name: String, factory: org.gradle.api.NamedDomainObjectFactory<T>): T = factory.create(name)
+
+            fun <T : Any> createK(type: kotlin.reflect.KClass<T>, factory: org.gradle.api.NamedDomainObjectFactory<T>): T = factory.create(type.simpleName!!)
+
+            fun test() {
+
+                // Implicit SAM conversion in regular source
+                println(createK(String::class) { it.toUpperCase() })
+                println(create("FOO") { it.toLowerCase() })
+
+                // Implicit SAM with receiver conversion in regular source
+                applyActionTo("BAR") {
+                    println(toLowerCase())
+                }
+            }
+        """)
+
+        withBuildScript("""
+
+            task("test") {
+                doLast { my.test() }
+            }
+
+         """)
+
+        assertThat(
+            build("test", "-q").output.also(::println),
+            containsMultiLineString("""
+                STRING
+                foo
+                bar
+            """)
+        )
     }
 
     private

--- a/subprojects/provider/src/main/kotlin/org/gradle/kotlin/dsl/KotlinBuildScript.kt
+++ b/subprojects/provider/src/main/kotlin/org/gradle/kotlin/dsl/KotlinBuildScript.kt
@@ -45,7 +45,12 @@ annotation class KotlinScriptTemplate
 @ScriptTemplateDefinition(
     resolver = KotlinBuildScriptDependenciesResolver::class,
     scriptFilePattern = ".*\\.gradle\\.kts")
-@ScriptTemplateAdditionalCompilerArguments(["-Xjsr305=strict"])
+@ScriptTemplateAdditionalCompilerArguments([
+    "-Xjsr305=strict",
+    "-XXLanguage:+NewInference",
+    "-XXLanguage:+SamConversionForKotlinFunctions",
+    "-Xprogressive"
+])
 @SamWithReceiverAnnotations("org.gradle.api.HasImplicitReceiver")
 @GradleDsl
 abstract class KotlinBuildScript(

--- a/subprojects/provider/src/main/kotlin/org/gradle/kotlin/dsl/KotlinInitScript.kt
+++ b/subprojects/provider/src/main/kotlin/org/gradle/kotlin/dsl/KotlinInitScript.kt
@@ -59,7 +59,12 @@ import kotlin.script.templates.ScriptTemplateDefinition
 @ScriptTemplateDefinition(
     resolver = KotlinBuildScriptDependenciesResolver::class,
     scriptFilePattern = ".+\\.init\\.gradle\\.kts")
-@ScriptTemplateAdditionalCompilerArguments(["-Xjsr305=strict"])
+@ScriptTemplateAdditionalCompilerArguments([
+    "-Xjsr305=strict",
+    "-XXLanguage:+NewInference",
+    "-XXLanguage:+SamConversionForKotlinFunctions",
+    "-Xprogressive"
+])
 @SamWithReceiverAnnotations("org.gradle.api.HasImplicitReceiver")
 abstract class KotlinInitScript(
     private val host: KotlinScriptHost<Gradle>

--- a/subprojects/provider/src/main/kotlin/org/gradle/kotlin/dsl/KotlinSettingsScript.kt
+++ b/subprojects/provider/src/main/kotlin/org/gradle/kotlin/dsl/KotlinSettingsScript.kt
@@ -69,7 +69,12 @@ import kotlin.script.templates.ScriptTemplateDefinition
 @ScriptTemplateDefinition(
     resolver = KotlinBuildScriptDependenciesResolver::class,
     scriptFilePattern = "^(settings|.+\\.settings)\\.gradle\\.kts$")
-@ScriptTemplateAdditionalCompilerArguments(["-Xjsr305=strict"])
+@ScriptTemplateAdditionalCompilerArguments([
+    "-Xjsr305=strict",
+    "-XXLanguage:+NewInference",
+    "-XXLanguage:+SamConversionForKotlinFunctions",
+    "-Xprogressive"
+])
 @SamWithReceiverAnnotations("org.gradle.api.HasImplicitReceiver")
 abstract class KotlinSettingsScript(
     private val host: KotlinScriptHost<Settings>

--- a/subprojects/provider/src/main/kotlin/org/gradle/kotlin/dsl/support/KotlinCompiler.kt
+++ b/subprojects/provider/src/main/kotlin/org/gradle/kotlin/dsl/support/KotlinCompiler.kt
@@ -43,6 +43,7 @@ import org.jetbrains.kotlin.config.JVMConfigurationKeys
 import org.jetbrains.kotlin.config.JVMConfigurationKeys.OUTPUT_DIRECTORY
 import org.jetbrains.kotlin.config.JVMConfigurationKeys.OUTPUT_JAR
 import org.jetbrains.kotlin.config.JVMConfigurationKeys.RETAIN_OUTPUT_IN_MEMORY
+import org.jetbrains.kotlin.config.LanguageFeature
 import org.jetbrains.kotlin.config.LanguageVersion
 import org.jetbrains.kotlin.config.LanguageVersionSettingsImpl
 import org.jetbrains.kotlin.config.addKotlinSourceRoot
@@ -209,6 +210,10 @@ private
 val languageVersionSettings = LanguageVersionSettingsImpl(
     languageVersion = LanguageVersion.KOTLIN_1_2,
     apiVersion = ApiVersion.KOTLIN_1_2,
+    specificFeatures = mapOf(
+        LanguageFeature.NewInference to LanguageFeature.State.ENABLED,
+        LanguageFeature.SamConversionForKotlinFunctions to LanguageFeature.State.ENABLED
+    ),
     analysisFlags = mapOf(
         AnalysisFlag.skipMetadataVersionCheck to true
     )

--- a/subprojects/provider/src/test/kotlin/org/gradle/kotlin/dsl/integration/KotlinBuildScriptIntegrationTest.kt
+++ b/subprojects/provider/src/test/kotlin/org/gradle/kotlin/dsl/integration/KotlinBuildScriptIntegrationTest.kt
@@ -1,6 +1,7 @@
 package org.gradle.kotlin.dsl.integration
 
 import org.gradle.kotlin.dsl.fixtures.AbstractIntegrationTest
+import org.gradle.kotlin.dsl.fixtures.containsMultiLineString
 
 import org.hamcrest.CoreMatchers.containsString
 
@@ -53,26 +54,7 @@ class KotlinBuildScriptIntegrationTest : AbstractIntegrationTest() {
     @Test
     fun `accepts lambda as SAM argument to Kotlin function`() {
 
-        withDefaultSettingsIn("buildSrc")
-
-        withBuildScriptIn("buildSrc", """
-
-            import org.jetbrains.kotlin.gradle.tasks.KotlinCompile
-
-            plugins {
-                `kotlin-dsl`
-            }
-
-            tasks.withType<KotlinCompile> {
-                kotlinOptions {
-                    freeCompilerArgs += listOf(
-                        "-XXLanguage:+NewInference",
-                        "-XXLanguage:+SamConversionForKotlinFunctions"
-                    )
-                }
-            }
-
-        """)
+        withKotlinBuildSrc()
 
         withFile("buildSrc/src/main/kotlin/my.kt", """
             package my
@@ -81,19 +63,7 @@ class KotlinBuildScriptIntegrationTest : AbstractIntegrationTest() {
 
             fun <T> create(name: String, factory: org.gradle.api.NamedDomainObjectFactory<T>): T = factory.create(name)
 
-            fun <T : Any> createK(type: kotlin.reflect.KClass<T>, factory: org.gradle.api.NamedDomainObjectFactory<T>): T = factory.create(type.simpleName!!)
-
-            fun test() {
-
-                // Implicit SAM conversion in regular source
-                println(createK(String::class) { it.toUpperCase() })
-                println(create("FOO") { it.toLowerCase() })
-
-                // Implicit SAM with receiver conversion in regular source
-                applyActionTo("BAR") {
-                    println(toLowerCase())
-                }
-            }
+            fun <T : Any> create(type: kotlin.reflect.KClass<T>, factory: org.gradle.api.NamedDomainObjectFactory<T>): T = factory.create(type.simpleName!!)
         """)
 
         withBuildScript("""
@@ -105,22 +75,29 @@ class KotlinBuildScriptIntegrationTest : AbstractIntegrationTest() {
                     // Explicit SAM conversion
                     println(create("foo", NamedDomainObjectFactory<String> { it.toUpperCase() }))
                     // Explicit SAM conversion with generic type argument inference
-                    println(create<String>("foo", NamedDomainObjectFactory { it.toUpperCase() }))
+                    println(create<String>("bar", NamedDomainObjectFactory { it.toUpperCase() }))
                     // Implicit SAM conversion
-                    println(create<String>("foo") { it.toUpperCase() })
-                    println(createK(String::class) { it.toUpperCase() })
-                    println(createK(String::class, { name: String -> name.toUpperCase() }))
+                    println(create<String>("baz") { it.toUpperCase() })
+                    println(create(String::class) { it.toUpperCase() })
+                    println(create(String::class, { name: String -> name.toUpperCase() }))
                     // Implicit SAM with receiver conversion
                     applyActionTo("action") {
                         println(toUpperCase())
                     }
-                    test()
                 }
             }
          """.replaceIndent())
 
         assertThat(
-            build("test", "-q").output.also(::println),
-            containsString("ACTION"))
+            build("test", "-q").output,
+            containsMultiLineString("""
+                FOO
+                BAR
+                BAZ
+                STRING
+                STRING
+                ACTION
+            """)
+        )
     }
 }

--- a/subprojects/provider/src/test/kotlin/org/gradle/kotlin/dsl/integration/KotlinBuildScriptIntegrationTest.kt
+++ b/subprojects/provider/src/test/kotlin/org/gradle/kotlin/dsl/integration/KotlinBuildScriptIntegrationTest.kt
@@ -49,4 +49,78 @@ class KotlinBuildScriptIntegrationTest : AbstractIntegrationTest() {
             assertThat(output, containsString("The plugins {} block must not be used here"))
         }
     }
+
+    @Test
+    fun `accepts lambda as SAM argument to Kotlin function`() {
+
+        withDefaultSettingsIn("buildSrc")
+
+        withBuildScriptIn("buildSrc", """
+
+            import org.jetbrains.kotlin.gradle.tasks.KotlinCompile
+
+            plugins {
+                `kotlin-dsl`
+            }
+
+            tasks.withType<KotlinCompile> {
+                kotlinOptions {
+                    freeCompilerArgs += listOf(
+                        "-XXLanguage:+NewInference",
+                        "-XXLanguage:+SamConversionForKotlinFunctions"
+                    )
+                }
+            }
+
+        """)
+
+        withFile("buildSrc/src/main/kotlin/my.kt", """
+            package my
+
+            fun <T> applyActionTo(value: T, action: org.gradle.api.Action<T>) = action.execute(value)
+
+            fun <T> create(name: String, factory: org.gradle.api.NamedDomainObjectFactory<T>): T = factory.create(name)
+
+            fun <T : Any> createK(type: kotlin.reflect.KClass<T>, factory: org.gradle.api.NamedDomainObjectFactory<T>): T = factory.create(type.simpleName!!)
+
+            fun test() {
+
+                // Implicit SAM conversion in regular source
+                println(createK(String::class) { it.toUpperCase() })
+                println(create("FOO") { it.toLowerCase() })
+
+                // Implicit SAM with receiver conversion in regular source
+                applyActionTo("BAR") {
+                    println(toLowerCase())
+                }
+            }
+        """)
+
+        withBuildScript("""
+
+            import my.*
+
+            task("test") {
+                doLast {
+                    // Explicit SAM conversion
+                    println(create("foo", NamedDomainObjectFactory<String> { it.toUpperCase() }))
+                    // Explicit SAM conversion with generic type argument inference
+                    println(create<String>("foo", NamedDomainObjectFactory { it.toUpperCase() }))
+                    // Implicit SAM conversion
+                    println(create<String>("foo") { it.toUpperCase() })
+                    println(createK(String::class) { it.toUpperCase() })
+                    println(createK(String::class, { name: String -> name.toUpperCase() }))
+                    // Implicit SAM with receiver conversion
+                    applyActionTo("action") {
+                        println(toUpperCase())
+                    }
+                    test()
+                }
+            }
+         """.replaceIndent())
+
+        assertThat(
+            build("test", "-q").output.also(::println),
+            containsString("ACTION"))
+    }
 }

